dialectic-compiler/src/lib.rs

```diff
@@ -0,0 +1,909 @@
+use {
+    lazy_static::lazy_static,
+    proc_macro2::{Span, TokenStream},
+    quote::{quote, ToTokens},
+    std::{fmt, ops, rc::Rc},
+    syn::{spanned::Spanned, Error, Ident, Type},
+    thiserror::Error,
+    thunderdome::{Arena, Index},
+};
+
+pub mod parse;
+
+#[derive(Error, Debug, Clone)]
+pub enum CompileError {
+    #[error("label name `'{0}` shadows a label name that is already in scope")]
+    ShadowedLabel(String),
+    #[error("undeclared label `'{0}`")]
+    UndeclaredLabel(String),
+    #[error("cannot `continue` outside of a loop")]
+    ContinueOutsideLoop,
+    #[error("cannot `break` outside of a loop")]
+    BreakOutsideLoop,
+}
+
+#[derive(Debug, Clone)]
+pub struct SpannedCompileError {
```

Comment:
There are many structures defined here that pair a `Span` with something else. Perhaps we can refactor all of them to use a shared type such as:

```rust
pub struct Spanned<T> {
    pub inner: T,
    pub span: Span,
}
```

Reply:
That's a good point. I like that; I'll swap to that in the commit that addresses these.